qiita_db/software.py

```diff
@@ -537,29 +609,214 @@ def command(self):
+
+            for key in cmd_opt_params:
+                try:
+                    values[key] = parameters.pop(key)
+                except KeyError:
+                    raise qdb.exceptions.QiitaDBError(
+                        "%s. Missing optional parameter: %s"
+                        % (error_msg, key))
+
+            if parameters:
+                raise qdb.exceptions.QiitaDBError(
+                    "%s. Extra parameters: %s"
+                    % (error_msg, ', '.join(parameters.keys())))
 
-            values = self.values
+            return cls(values, command)
+
+    @classmethod
+    def from_default_params(cls, dflt_params, req_params, opt_params=None):
+        """Creates the parameter set from a `dflt_params` set
+
+        Parameters
+        ----------
+        dflt_params : qiita_db.software.DefaultParameters
+            The DefaultParameters object in which this instance is based on
+        req_params : dict of {str: object}
+            The required parameters values, keyed by parameter name
+        opt_params : dict of {str: object}, optional
+            The optional parameters to change from the default set, keyed by
+            parameter name. Default: None, use the values in `dflt_params`
 
-            result = []
-            for p_name, p_type in sorted(table_cols):
-                if p_type == 'boolean':
-                    if values[p_name]:
-                        result.append("--%s" % p_name)
-                else:
-                    result.append("--%s '%s'" % (p_name, values[p_name]))
+        Raises
+        ------
+        QiitaDBError
+            - If there are missing requried parameters
+            - If there is an unknown required ot optional parameter
+        """
+        with qdb.sql_connection.TRN:
+            command = dflt_params.command
+            cmd_req_params = command.required_parameters
+            cmd_opt_params = command.optional_parameters
+
+            missing_reqd = set(cmd_req_params) - set(req_params)
+            extra_reqd = set(req_params) - set(cmd_req_params)
+            if missing_reqd or extra_reqd:
+                raise qdb.exceptions.QiitaDBError(
+                    "Provided required parameters not expected.\n"
+                    "Missing required parameters: %s\n"
+                    "Extra required paramters: %s\n"
+                    % (', '.join(missing_reqd), ', '.join(extra_reqd)))
+
+            if opt_params:
+                extra_opts = set(opt_params) - set(cmd_opt_params)
+                if extra_opts:
+                    raise qdb.exceptions.QiitaDBError(
+                        "Extra optional parameters provded: %s"
+                        % ', '.join(extra_opts))
+
+            values = dflt_params.values
+            values.update(req_params)
+
+            if opt_params:
+                values.update(opt_params)
+
+            return cls(values, command)
+
+    def __init__(self, values, command):
```

Comment:
Usually the numpy docstring of the `__init__` method goes to the class level documentation, which I intentionally left it out except for the Raises section since I don't want users to use the __init__ method (they should go through the classmethods)